core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -720,6 +720,29 @@ public Endpoint withAttrs(Attributes newAttributes) {
         return new Endpoint(type, host, ipAddr, port, weight, newAttributes);
     }
 
+    /**
+     * Returns a new {@link Endpoint} with the specified {@link Attributes}.
+     * Note that the {@link #attrs()} of this {@link Endpoint} is merged with the specified
+     * {@link Attributes}. For attributes with the same {@link AttributeKey}, the attribute
+     * in {@param newAttributes} has higher precedence.
+     */
+    @SuppressWarnings("unchecked")
+    public Endpoint mergeAttrs(Attributes newAttributes) {
```

Comment:
Maybe unrelated, but I noticed that the current `withAttrs` behavior is somewhat confusing. `withAttr(AttributeKey, ...)` doesn't replace the entire attributes but `withAttrs()` does, and yet they both use `with` prefix. I feel like we need this change first:

- Change the semantics of `withAttrs()` so it doesn't replace all attributes.
- Introduce `replaceAttrs()` that provides the original behavior.

Reply:
Addressed this, but note that this PR will act as an umbrella PR to let reviewers know the overall direction of the health check changes.
This PR will be split into multiple PRs for more thorough/readable changes as usual